### PR TITLE
mgr/volumes: Retain suid guid bits in clone

### PR DESCRIPTION
**Pull request by kotreshhr**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by kotreshhr** [comment]
jenkins test make check

File: src/include/cephfs/libcephfs.h [hunk 1]
```diff
@@ -883,6 +883,17 @@ int ceph_chmod(struct ceph_mount_info *cmount, const char *path, mode_t mode);
...
+
+/**
+ * Change the mode bits (permissions) of a file/directory. If the path is a
+ * symbolic link, it's not de-referenced.
+ *
+ * @param cmount the ceph mount handle to use for performing the chmod.
+ * @param path the path of file/directory to change the mode bits on.
+ * @param mode the new permissions to set.
+ * @returns 0 on success or a negative error code on failure.
+ */
```
Comment (batrick) [inline src/include/cephfs/libcephfs.h:896]: docs go above the function?

File: src/include/cephfs/libcephfs.h [hunk 2]
```diff
@@ -883,6 +883,17 @@ int ceph_chmod(struct ceph_mount_info *cmount, const char *path, mode_t mode);
  * @param mode the new permissions to set.
  * @returns 0 on success or a negative error code on failure.
  */
+int ceph_lchmod(struct ceph_mount_info *cmount, const char *path, mode_t mode);
```
Comment (batrick) [inline src/include/cephfs/libcephfs.h:886]: This is stealing the docs for `fchmod` below.

File: src/test/libcephfs/test.cc [hunk 3]
```diff
@@ -814,6 +814,46 @@ TEST(LibCephFS, Fchmod) {
...
+
+  // Change mode on symlink file
+  ASSERT_EQ(ceph_lchmod(cmount, test_symlink, 0400), 0);
+  struct ceph_statx stx_orig2;
+  ASSERT_EQ(ceph_statx(cmount, test_symlink, &stx_orig2, CEPH_STATX_ALL_STATS, AT_SYMLINK_NOFOLLOW), 0);
+
+  // Compare modes
+  ASSERT_NE(stx_orig1.stx_mode, stx_orig2.stx_mode);
+  ASSERT_EQ(stx_orig1.stx_mode, 0120777);
+  ASSERT_EQ(stx_orig2.stx_mode, 0120400);
```
Comment (batrick) [inline src/test/libcephfs/test.cc:852]:
```suggestion
Where did you get the `0120XXX` magic number? I think the POSIXy correct test is:

```suggestion
  static const int permbits = S_IRWXU|S_IRWXG|S_IRWXO;
  ASSERT_EQ(permbits&stx_orig1.stx_mode, 0777);
  ASSERT_EQ(permbits&stx_orig2.stx_mode, 0400);
```

per https://pubs.opengroup.org/onlinepubs/009604499/basedefs/sys/stat.h.html

>  The file permission bits are defined to be those corresponding to the bitwise-inclusive OR of S_IRWXU, S_IRWXG, and S_IRWXO.

File: src/test/pybind/test_cephfs.py [hunk 4]
```diff
@@ -540,6 +540,30 @@ def test_futimens():
...
+
+    time.sleep(1)
+    cephfs.lchmod(b'/file-2', 0o400)
+
+    stx_post_t = cephfs.statx(b'/file-1', libcephfs.CEPH_STATX_MODE, 0)
+    stx_post_s = cephfs.statx(b'/file-2', libcephfs.CEPH_STATX_MODE, libcephfs.AT_SYMLINK_NOFOLLOW)
+
+    assert_equal(stx_post_t['mode'], stx_pre_t['mode'])
+    assert_not_equal(stx_post_s['mode'], stx_pre_s['mode'])
+    assert_equal(stx_post_s['mode'], 0o120400)
```
Comment (batrick) [inline src/test/pybind/test_cephfs.py:562]: ditto, but here you can assume `S_IRWXU|S_IRWXG|S_IRWXO` is `0777`.

**Review by batrick — CHANGES_REQUESTED** [review]
otherwise lgtm; great work!

File: src/test/libcephfs/test.cc (hunk 3 above)
Comment (kotreshhr) [reply src/test/libcephfs/test.cc:852]: That number is from testing, it includes file type as well. This is more readable. I will use this.

Comment (kotreshhr) [reply src/test/libcephfs/test.cc]: Done

File: src/test/pybind/test_cephfs.py (hunk 4 above)
Comment (kotreshhr) [reply src/test/pybind/test_cephfs.py]: Done

**Review by batrick — APPROVED** [review]
Please run through testing.

**Comment by ajarr** [comment]
teuthology testing with --filter volumes,libcephfs_python,tasks/libcephfs
http://pulpito.front.sepia.ceph.com/rraja-2021-03-18_22:56:35-fs-wip-khiremat-40214-retain-suid-subvol-clone-distro-basic-smithi/

**Review by ajarr — APPROVED** [review]

**Comment by ajarr** [comment]
> teuthology testing with --filter volumes,libcephfs_python,tasks/libcephfs
> http://pulpito.front.sepia.ceph.com/rraja-2021-03-18_22:56:35-fs-wip-khiremat-40214-retain-suid-subvol-clone-distro-basic-smithi/

Test failures seem unrelated.
I have not seen this before,
http://pulpito.front.sepia.ceph.com/rraja-2021-03-18_22:56:35-fs-wip-khiremat-40214-retain-suid-subvol-clone-distro-basic-smithi/5978421/
newly added lchmod test passes, but symlinks test fails

**Comment by kotreshhr** [comment]
> > teuthology testing with --filter volumes,libcephfs_python,tasks/libcephfs
> > http://pulpito.front.sepia.ceph.com/rraja-2021-03-18_22:56:35-fs-wip-khiremat-40214-retain-suid-subvol-clone-distro-basic-smithi/
> 
> Test failures seem unrelated.
> I have not seen this before,
> http://pulpito.front.sepia.ceph.com/rraja-2021-03-18_22:56:35-fs-wip-khiremat-40214-retain-suid-subvol-clone-distro-basic-smithi/5978421/
> newly added lchmod test passes, but symlinks test fails

I think it runs as single process and the symlink file names are conflicting between 'Lchmod' and "Symlinks" tests. I will change and refresh this PR.

**Comment by kotreshhr** [comment]
jenkins test api

**Comment by kotreshhr** [comment]
Teuthology results:

http://pulpito.front.sepia.ceph.com/khiremat-2021-03-19_05:25:43-fs-wip-khiremat-40214-retain-suid-subvol-clone-1-distro-basic-smithi/

The symlink testcase is fixed, but few unrelated failures.

**Comment by kotreshhr** [comment]
jenkins test make check

**Comment by kotreshhr** [comment]
jenkins test api

**Comment by batrick** [comment]
LGTM

**Pull request merged**